src/main/kotlin/org/rust/cargo/project/model/impl/CargoProjectImpl.kt

```diff
@@ -0,0 +1,378 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.cargo.project.model.impl
+
+import com.intellij.execution.ExecutionException
+import com.intellij.execution.process.ProcessAdapter
+import com.intellij.execution.process.ProcessEvent
+import com.intellij.notification.NotificationType
+import com.intellij.openapi.application.ApplicationManager
+import com.intellij.openapi.application.runWriteAction
+import com.intellij.openapi.components.PersistentStateComponent
+import com.intellij.openapi.components.State
+import com.intellij.openapi.diagnostic.Logger
+import com.intellij.openapi.progress.BackgroundTaskQueue
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.roots.ModuleRootManager
+import com.intellij.openapi.roots.ex.ProjectRootManagerEx
+import com.intellij.openapi.util.EmptyRunnable
+import com.intellij.openapi.util.Key
+import com.intellij.openapi.vfs.LocalFileSystem
+import com.intellij.openapi.vfs.VirtualFile
+import com.intellij.openapi.vfs.VirtualFileManager
+import com.intellij.util.Consumer
+import com.intellij.util.indexing.LightDirectoryIndex
+import com.intellij.util.io.exists
+import com.intellij.util.io.systemIndependentPath
+import org.jdom.Element
+import org.jetbrains.annotations.TestOnly
+import org.jetbrains.concurrency.Promise
+import org.jetbrains.concurrency.collectResults
+import org.rust.cargo.project.model.CargoProject
+import org.rust.cargo.project.model.CargoProject.UpdateStatus
+import org.rust.cargo.project.model.CargoProjectsService
+import org.rust.cargo.project.settings.RustProjectSettingsService
+import org.rust.cargo.project.settings.rustSettings
+import org.rust.cargo.project.settings.toolchain
+import org.rust.cargo.project.workspace.CargoWorkspace
+import org.rust.cargo.project.workspace.PackageOrigin
+import org.rust.cargo.project.workspace.StandardLibrary
+import org.rust.cargo.toolchain.RustToolchain
+import org.rust.cargo.toolchain.Rustup
+import org.rust.cargo.util.modules
+import org.rust.ide.notifications.showBalloon
+import org.rust.utils.AsyncValue
+import org.rust.utils.TaskResult
+import org.rust.utils.pathAsPath
+import org.rust.utils.runAsyncTask
+import java.nio.file.Path
+import java.nio.file.Paths
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.atomic.AtomicReference
+
+private val LOG = Logger.getInstance(CargoProjectsServiceImpl::class.java)
+
+@State(name = "CargoProjects")
+class CargoProjectsServiceImpl(
```

Comment:
I think testing of this service would be cool

Reply:
An excellent suggestion @farodin91 !

I've added some additional test for `AsyncValue`, and looks like `Promise` has a bug, and we should switch to `CompletableFuture`!